### ISSUE TITLE
Slow parsing due to huge ClassImpl.subClasses in GarbageCleaner

### ISSUE THREAD

**Issue by turbanoff** [eclipse-mat/mat#59]
I have a heap dump, which Eclipse MAT have troubles to open.
![image](https://github.com/eclipse-mat/mat/assets/741251/ce8da468-621f-4d5c-96ad-2a76829b9e55)

Thread dump always show this stack trace:
```
"Worker-1: Parsing heap dump from 'C:\Users\turbanov\Documents\broqa_dumps\java_pid789.hprof\java_pid789.hprof'@8474" prio=5 tid=0x29 nid=NA runnable
  java.lang.Thread.State: RUNNABLE
	  at org.eclipse.mat.parser.model.ClassImpl.removeSubClass(ClassImpl.java:467)
	  at org.eclipse.mat.parser.internal.GarbageCleaner.clean(GarbageCleaner.java:238)
	  at org.eclipse.mat.parser.internal.SnapshotFactoryImpl.parse(SnapshotFactoryImpl.java:466)
	  at org.eclipse.mat.parser.internal.SnapshotFactoryImpl.openSnapshot(SnapshotFactoryImpl.java:224)
	  at org.eclipse.mat.snapshot.SnapshotFactory.openSnapshot(SnapshotFactory.java:149)
	  at org.eclipse.mat.ui.snapshot.ParseHeapDumpJob.run(ParseHeapDumpJob.java:99)
	  at org.eclipse.core.internal.jobs.Worker.run(Worker.java:63)
```

I was able to attach debugger.
Seems issue with huge lists, used by `org.eclipse.mat.parser.internal.GarbageCleaner`
First one is - `org.eclipse.mat.parser.model.ClassImpl#subClasses`
![image](https://github.com/eclipse-mat/mat/assets/741251/378b9f1e-b5e6-49ea-911c-5f6f36bedbec)
Second one is - `classes2remove`
![image](https://github.com/eclipse-mat/mat/assets/741251/be9de62a-5b4c-4665-8a98-b70cb6a84c94)

I think at least `org.eclipse.mat.parser.model.ClassImpl#subClasses` could be converted to HashSet/LinkedHashSet to make `ClassImpl.removeSubClass` to be run in O(1).

**Comment by turbanoff** [comment]
Yes, solution to use LinkedHashSet instead of ArrayList for `ClassImpl.subClasses` works. After changing I was able to open the heap dump in reasonable time.

**Comment by jasonk000** [comment]
looks related in #57

**Comment by jasonk000** [comment]
@turbanoff , thank you for the issue/analysis/fix. Are you able to share the heapdump or do any more analysis here?

As we saw in #57, the patch to GarbageCleaner to use FJ pool seems to have triggered this.
https://github.com/eclipse-mat/mat/commit/7c2cb094fc6119999e14e1d8a04fe15c37c31f34

But the patch does not change the ClassImpl, which makes me wonder -- does the new `GarbageCleaner.clean()` implementation have a different bug that is leading to a larger list than expected?

**Comment by turbanoff** [comment]
Here the dump, which Eclipse MAT opens very slowly. Was able to create it with quite small program.
[java_pid12932.hprof.gz](https://github.com/user-attachments/files/15820719/java_pid12932.hprof.gz)

**Comment by jasonk000** [comment]
@turbanoff thanks for the report, fix, and repro, I've put together patch in #63 , though I only addressed the first list/set, which led to a significant boost. Let me know if you disagree with it.
